src/TableOfContents/index.tsx

```diff
@@ -249,148 +207,120 @@ export function TableOfContents<T extends TableOfContentsItem = TableOfContentsI
 
   let icon = item.icon;
-  if (isActive || isSelected) {
-    if (item.activeIcon) {
-      icon = item.activeIcon;
-    }
-
-    if (forceStateStyle === 'active') {
-      isActive = true;
-      isSelected = false;
-    } else if (forceStateStyle === 'selected') {
-      isActive = false;
-      isSelected = true;
-    }
+  if (item.activeIcon && (isActive || isSelected)) {
+    icon = item.activeIcon;
   }
 
-  if (showSkeleton) {
-    isActive = false;
-    isSelected = false;
-  }
+  const onClick = showSkeleton
+    ? undefined
+    : (e: React.MouseEvent) => {
+        if (item.isDisabled) {
+          e.preventDefault();
+          return;
+        }
+        if (item.onClick) {
+          item.onClick();
+        }
+
+        if (isDivider) {
+          e.preventDefault();
+          return;
+        }
 
-  const className = cn(
+        if (!isGroup) return;
+
+        e.preventDefault();
+        toggleExpanded();
+      };
+
+  const outerClassName = cn('TableOfContentsItem border-transparent', item.className, {
+    'border-l': !isGroup,
+    'TableOfContentsItem--selected': isActive,
+    'TableOfContentsItem--active': isSelected,
+    'TableOfContentsItem--group': isGroup,
+    'TableOfContentsItem--divider': isDivider,
+    'TableOfContentsItem--child border-gray-3 dark:border-lighten-3': isChild,
+  });
+
+  const innerClassName = cn(
     'TableOfContentsItem__inner relative flex flex-col justify-center border-transparent border-l-4',
     {
-      'cursor-pointer': (item.onClick || onClick) && !showSkeleton && !isDisabled,
+      'cursor-pointer': onClick && !isDisabled,
       'cursor-not-allowed': isDisabled,
       'dark-hover:bg-lighten-2 hover:bg-darken-2':
         !isDisabled && !isDivider && !isSelected && !isActive && !showSkeleton,
       'dark:text-white bg-darken-2 dark:bg-lighten-2': isSelected || isActive,
       'text-gray-7 dark:text-white': isActive,
       'border-primary text-blue-6': isSelected,
-
       'text-gray-6 dark:text-gray-6 font-semibold h-10': isDivider,
       'text-gray-5 dark:text-gray-5 hover:text-gray-6': !isDivider && !isSelected && !isActive,
     },
   );
 
-  let loadingElem;
-  if (item.isLoading) {
-    loadingElem = <FAIcon icon={['far', 'spinner-third']} className="fa-spin text-gray-7 ml-2" />;
-  }
-
-  let actionElem;
-  if (item.action) {
-    actionElem = (
-      <Button
-        icon={
-          item.action.icon ? (
-            <FAIcon icon={item.action.icon} className={cn({ 'text-gray-5': !item.action.isActive })} />
-          ) : undefined
-        }
-        text={item.action.name}
-        onClick={showSkeleton ? undefined : item.action.onClick}
-        active={item.action.isActive}
-        intent={item.action.isActive ? 'primary' : undefined}
-        className="ml-2"
-        minimal
-        small
-      />
-    );
-  }
+  const loadingElem = item.isLoading ? (
+    <FAIcon icon={['far', 'spinner-third']} className="fa-spin text-gray-7 ml-2" />
+  ) : null;
```

Comment:
I usually do and I try to avoid `null` as much as possible, but in case of JSX I prefer `null`. 

This is because in the React-JSX ecosystem `null` means an explicit _don't render anything_ while `undefined` can be problematic:

If the containing component did
```
return loadingElem;
```

This would be valid. If I changed the definition to `void 0` the above would trigger an error.

Reply:
That was my thought but I don't have a strong opinion on these. Would you prefer if I changed it to `void 0`?

Reply:
No, I actually think `null` is better here. That was just unrelated comment :)